### ISSUE TITLE
Next/Jest : Cannot override transformIgnorePatterns

### ISSUE THREAD

**Issue by crubier** [vercel/next.js#35634]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

Relevant in any environment

### Describe the Bug

The way `next/jest` is implemented, ([See here](https://github.com/vercel/next.js/blob/canary/packages/next/build/jest/jest.ts#L124)), there is no way to force jest to transpile modules inside node_modules.

When trying to import node_modules who need to be transpiled (e.g the openlayers module `ol`), jest throws the following error:

```
Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation
```

The [recommended way to fix this error](https://jestjs.io/docs/configuration#transformignorepatterns-arraystring) is to add a `transformIgnorePatterns` regex, so that the `jest.config.js` looks like this:

```
// jest.config.js
const nextJest = require('next/jest')

const createJestConfig = nextJest({
  dir: '.',
})

const customJestConfig = {
  testEnvironment: 'jest-environment-jsdom',
  // ******* THIS BIT IS IMPORTANT *******
  transformIgnorePatterns: [
    '.*/node_modules/(?!(ol)/)'
  ]
}

module.exports = createJestConfig(customJestConfig)
```

BUT this does not work due to the way `next/jest` is implemented, ([See here](https://github.com/vercel/next.js/blob/canary/packages/next/build/jest/jest.ts#L124)), the `node_module` element is impossible to remove from that array, and hence no `node_module` can be transpiled, ever, even if I override this in my own `jest.config.js` .

### Expected Behavior

Setting this in my own `jest.config.js`

```
  transformIgnorePatterns: [
    '.*/node_modules/(?!(ol)/)'
  ]
```

Would transpile `ol`, and I wouldn't have this error when I run tests:

```

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

```

### To Reproduce

1. Clone the official next jest example: https://github.com/vercel/next.js/tree/canary/examples/with-jest
2. `npm install ol`
3. Put this the `jest.config.js`:
```
// jest.config.js
const nextJest = require('next/jest')
const createJestConfig = nextJest({your test environment
  dir: '.',
})
const customJestConfig = {
  testEnvironment: 'jest-environment-jsdom',
  transformIgnorePatterns: [
    '.*/node_modules/(?!(ol)/)'
  ]
}
module.exports = createJestConfig(customJestConfig)
```
4. Create a test: `my.test.ts`:
```
// my.test.ts
import { Map } from "ol";
describe("Testing", () => {
  it("can run ol tests", () => {
    const map = new Map({});
    expect(map.getAllLayers().length).toBe(0);
  });
});
```
5. See the error.

**Comment by xveganxxxedgex** [comment]
I'm running into this issue as well for `moduleNameMapper` - I need to map `.svg` files to be stubbed as react components, but the existing image stub rule in the next/jest config is getting used instead since it's higher on the list:

https://github.com/vercel/next.js/blob/acbd54322f447caaee09b1262e4687434ea0268c/packages/next/build/jest/jest.ts#L74

**Comment by pzi** [comment]
Looks like @kevva was on the money here: https://github.com/vercel/next.js/pull/34350#issuecomment-1055463916

**Comment by kevva** [comment]
For your info, I solved it by mocking them manually in the test files instead:

```js
jest.mock('path/to/image.svg', () => 'svg');
```

However, it'd be nice if we could override/extend the default values in `moduleNameMapper` ourselves.

**Comment by pzi** [comment]
Yeah, thanks to this comment https://github.com/vercel/next.js/issues/26749#issuecomment-885431747, I got to the same workaround for now. Thanks @kevva

**Comment by pzi** [comment]
@kevva, did you end up creating an issue for the moduleNameMapper order?

**Comment by kevva** [comment]
No, unfortunately not. I don't have time for writing a good issue with a good reproduction case atm. Feel free to go ahead and do it :).

**Comment by timneutkens** [comment]
I've moved the moduleNameMapper so that you can override the mapping. The case where you want to do anything else than is provided by default is explained well by @xveganxxxedgex's comment: https://github.com/vercel/next.js/issues/35634#issuecomment-1080942525. That approach allows you to override any setting you'd want to tweak based on customizations you've done to the webpack config, those are not covered / supported by Next.js and as such are not specifically added as features to next/jest.

**Comment by timmywil** [comment]
The PR that closed this issue did not address OP's original use case, which was not related to `moduleNameMapper`. The PR addressed a separate issue presented in the first comment.

The problem specifically has to do with jest 28 adding support for the exports field and jsdom's default environment being the browser. When the test environment is set to jsdom, any node module that exports esm will cause this error, which will only get more common as more packages add support for exports.

It's not really a jest error, so I think the best way to address it is to allow for specific node modules to be transformed in the jest config, which means overriding `transformIgnorePatterns` somehow and not just extending it. Yes, there are workarounds, but they are only bandaids.

Minimal reproduction: https://codesandbox.io/s/magical-chihiro-xf68kl  Open a new terminal window and run `npm test`.

**Comment by Talendar** [comment]
I'm still experiencing this issue in Next v12.1.6 while using Jest v28.0.3. I'm currently using the solution proposed by @timmywil as a workaround.
